Allow creation of FunctionsOfTime with control system info

knelli2 opened the pull request:
## Proposed changes

Adds a second simple tag derived off the `domain::Tags::FunctionsOfTime` base tag. This new tag uses control system information (specifically control system OptionHolders) to construct the functions of time so it is placed in the control system library.

There are additional commits that 1) add a function to the Controller and 2) change the template of the control system OptionHolder to allow these changes.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Change template of control sys option holder

knelli2 committed:
Add get_update_fraction memeber to controller

knelli2 committed:
Add FoT tag initialized with control systems

This tag takes some `control_system::OptionHolder`s as option_tags
for its `create_from_options` member in addition to the domain
creator and initial time step. It is similar to the tag in
`domain::Tags::FunctionsOfTimeInitialize` except that this new tag
depends on the control system, so it must be located outside the
Domain library.

nilsvu commented on src/ControlSystem/Actions/Initialization.hpp:
```diff
@@ -32,9 +32,9 @@ namespace Actions {
  *   - `control_system::Tags::Averager<2>`
  *   - `control_system::Tags::Controller<2>`
  *   - `control_system::Tags::TimescaleTuner`
+ *   - `control_system::Tags::ControlSystemName`
```
Why do you have this tag in the box? It's compile-time, right, so can't you just do `ControlSystem::name()` everywhere?

knelli2 replied:
Doing `ControlSystem::name()` everywhere would require that any action you want to run on the ControlComponent be templated on the control system (in case you need the name). That's a fairly restrictive requirement. Adding a tag for the name in the box relaxes this requirement because as long as you have access to the box (which is more readily passed around) you have access to the name.

nilsvu replied:
Ok I think passing around the control system template might be the way to go, but let's keep things as they are for now.

wthrowe replied:
Other possibilities to consider later: get the control system from the parallel component or put the control system struct itself in the DataBox.

nilsvu commented on src/ControlSystem/Actions/Initialization.hpp:
```diff
@@ -71,7 +72,7 @@ struct Initialize {
     ::Initialization::mutate_assign<tags_to_be_initialized>(
         make_not_null(&box), option_holder.averager, option_holder.controller,
-        option_holder.tuner);
+        option_holder.tuner, option_holder.name);
```
If you need the name in the box (see comment above) then you can still use `ControlSystem::name` here instead of going through the options (since the name is not an option). You can keep the templating on `ControlSystem` (I think it makes it easier to understand code if templating is consistent. `ControlSystem` has a well-defined protocol, so passing it around as a template parameter is good. You can actually insert more `assert_conforms_to` to validate the template parameter.)

knelli2 replied:
Yes I agree in this instance I can use `ControlSystem::name()`

However, right now the name has to be in the OptionHolder because inside the `FunctionsOfTimeInitialize` tag, we don't pass control systems, we pass OptionHolders. Yet we still needed the name so the OptionHolder seemed a nice place to put this because it succinctly holds all info needed to initialize a control system and functions of time.

However, since the OptionHolder is now templated on a control system, I could add an alias in the OptionHolder `using control_system =` instead of the `name` memeber and then inside the `FunctionsOfTimeInitialize` tag do `std::string name = decltype(option_holder)::control_system::name()` (or something similar) using the static `name()` from the ControlSystem. What do you think?

nilsvu replied:
Yea that sounds good. When a class has a template parameter there's no reason not to expose it as a type alias.

nilsvu commented on src/ControlSystem/Tags.hpp:
```diff
@@ -109,6 +126,63 @@ struct Controller : db::SimpleTag {
+/// \ingroup ControlSystemGroup
+/// The FunctionsOfTime initialized from a DomainCreator, initial time
+/// step, and control system OptionHolders.
+struct FunctionsOfTimeInitialize : domain::Tags::FunctionsOfTime,
```
Please move this into a separate file, `ControlSystem/Tags/FunctionsOfTimeInitialize.hpp`

knelli2 replied:
For consistency should I also move the whole `Tags.hpp` file to `Tags/Tags.hpp` (while also moving `FunctionsOfTimeInitialize` to a new file)?

nilsvu replied:
No I'd leave it where it is for now. The individual tags may eventually migrate to separate files though.

nilsvu commented on src/ControlSystem/Tags.hpp:
```diff
@@ -109,6 +126,63 @@ struct Controller : db::SimpleTag {
+          domain_creator,
+      const double initial_time_step, const OptionHolders&... option_holders) {
+    std::unordered_map<std::string, double> initial_expiration_times{};
+    [[maybe_unused]] const auto lambda = [&initial_expiration_times,
```
Please give the lambda a more descriptive name

nilsvu commented on src/ControlSystem/Tags.hpp:
```diff
@@ -175,17 +249,18 @@ struct MeasurementTimescales : db::SimpleTag {
-template <size_t DerivOrder>
+template <typename ControlSystem>
 struct OptionHolder {
+  static constexpr size_t deriv_order = ControlSystem::deriv_order;
```
Here, for instance, would be a good place for an `assert_conforms_to`, before this line

nilsvu reviewed (dismissed):
LGTM please squash

nilsvu commented:
Thanks for finding a nice way to factor these changes out of #3612 with this PR 👍

wthrowe commented on tests/Unit/ControlSystem/Test_Tags.cpp:
```diff
@@ -12,27 +12,43 @@
 
 namespace {
-const double initial_time = 2.0;
+const double initial_time = 0.0;
```
How hard is it to keep this a nonzero value?  Assuming an initial time of zero somewhere would be a really easy error to make.

knelli2 replied:
Not hard. I changed it to keep the initial time step small in the test of the FunctionsOfTimeInitialize tag, but for testing it just has to be small enough to not interfere with the initial expiration times.. I'll revert this to 2.0 and change other things to accommodate.

knelli2 commented:
Squashed in previous changes. After making the changes @wthrowe requested about the initial time, I realized I had been assuming the initial time was 0.0 for most of this. I added a fixup commit where I also take `OptionTags::InitialTime` as an argument to `FunctionsOfTimeInitialize` to account for this.

wthrowe commented:
Squash.
